### PR TITLE
Use SearchRouter everywhere and drop Chat finder

### PR DESCRIPTION
**Pull request by Kicu**
### Details
 - this PR adds the new `SearchButton` everywhere, replacing old `ChatFinder` page
 - `<SearchRouter>` is not a separate Page/Screen but rather just a plain react component that is displayed in a modal and accessible from almost every page in the App
 - everything related to `ChatFinder` is removed, and two e2e tests are rewritten to test the new SearchRourter component
   - While SearchRouter is doing much more than old ChatFinder, the logic within router that searches rooms is directly copied from ChatFinder - so I hope old test cases and patterns still make sense
   - I have not added any more tests scenarios because the main thing search router is doing I don't expect to be very heavy; we might want to add some scenarios in future once Search v2.5 has autocomplete in place
 - please ignore failing `Changed files ESLint check` because I touched `AuthScreens` which should be migrated in a dedicated PR

### Fixed Issues
$ https://github.com/Expensify/App/issues/49123
PROPOSAL:

### Tests

 - verify that the icon to open Search is available on every page and opens router when pressed
 - verify that using shortcut `cmd + k` opens router
 - check that e2e tests still make sense

### QA Steps

 - verify that the icon to open Search is available on every page and opens router when pressed
 - verify that using shortcut `cmd + k` opens router

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/6946e67c-842f-4d7d-81a9-87c712625bc2

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/4db8713b-bcf4-4c3c-bade-f79c7bc1560f
</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/bf35c48f-c8fb-403d-876a-0374baa5179d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/d378af83-e032-4c97-aa1d-b5be6dffe218

https://github.com/user-attachments/assets/14735d87-e892-489c-9ebf-0662a804f44a

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by Kicu** [comment]
@luacmartins @rayane-djouah This PR enables it all: takes off the `Permission` check for isDev and turns on SearchRouter everywhere including on cmd+k shortcut. However I'm going OOO for the next week and will be back 14 October.

Meanwhile if there are problems or deploy blockers then @SzymczakJ should be able to help with those.

I understand that we might notice some more bugs since probably now it will be tested more heavily 😅

**Comment by Kicu** [comment]
CC @adhorodyski if you mind taking a look at the file `tests/perf-test/SearchRouter.perf-test.tsx` in which there are old file `ChatFinder` reassure tests rewritten to new component. The tests were passing on my local dev setup, but I'd like to now if they make sense :)

**Comment by luacmartins** [comment]
Nice! Really excited about this one.

**Review by luacmartins — CHANGES_REQUESTED** [review]
BUG: Pressing `cmd + k` with the modal open closes the modal and reopens it

1. Press `cmd + k` to open the modal
2. With the modal open, press `cmd + k` again
3. Note that the modal closes and reopens

Expected behavior: the modal should just close if it is already open

**Comment by luacmartins** [comment]
@Expensify/design  I kicked off an adhoc build and would love your eyes on this one.

**Comment by luacmartins** [comment]
Should the modal have the same top margin as the search bar?
Screenshot 2024-10-04 at 9 02 59 AM

**Comment by Kicu** [comment]
> BUG: Pressing cmd + k 

I'll push a quick fix for this

**Comment by Kicu** [comment]
pushed a fix but its a bit dirty, not sure if it will hold up over review 😅 that's the best I could do for today to make it behave somewhat good.
The caveat is we will now not close other modals globally.

**Comment by SzymczakJ** [comment]
I'll be finishing this PR when Mateusz is OOO. As for:

> The caveat is we will now not close other modals globally.

I fixed it and now modals will close on cmd+k
Also, could you review tests section in this PR @adhorodyski?. Your opinion would be invaluable for us 😃

**Comment by adhorodyski** [comment]
On it now!

File: tests/perf-test/SearchRouter.perf-test.tsx
Comment (adhorodyski) [inline tests/perf-test/SearchRouter.perf-test.tsx:162]: exactly what we aimed for here 👍🏼

File: tests/perf-test/SearchRouter.perf-test.tsx
Comment (adhorodyski) [inline tests/perf-test/SearchRouter.perf-test.tsx:1]: thanks for removing some of the mocks!

**Review by adhorodyski — APPROVED** [review]

**Comment by dubielzyk-expensify** [comment]
This is looking awesome. Here's a few design fixes:

- The height of the recent searches should be `52px`. In the branch it's `44px`
- The section titles ("Recent searches", "Recent chats") should have 12px padding instead of 20px
- Router should be `6px` padding from the edge, not `20px`

Below a few screens to explain further:
CleanShot 2024-10-08 at 13 52 18@2x

CleanShot 2024-10-08 at 13 56 29@2x

CleanShot 2024-10-08 at 13 58 45@2x

**Comment by SzymczakJ** [comment]
> Search field should have our normal text size 16px

Did you mean 15px @dubielzyk-expensify? I see that `variables.fontSizeNormal` in our codebase are defined with `15px` size and also Figma mock has 15px text in it.

**Comment by SzymczakJ** [comment]
this is ready for your review @rayane-djouah 🙏

**Comment by dannymcclain** [comment]
>Did you mean 15px @dubielzyk-expensify? I see that variables.fontSizeNormal in our codebase are defined with 15px size and also Figma mock has 15px text in it.

Yes `15px` is correct 🙂

**Comment by dubielzyk-expensify** [comment]
My bad for the typo. Danny is correct 🙏

**Comment by dannymcclain** [comment]
>Do we need to add a margin between the search button and the "new workspace" button?
>I think we need to add a 16px gap like we did in search page

Yeah I agree.

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (dubielzyk-expensify) [reply src/components/Search/SearchRouter/SearchRouter.tsx:157]: Yep. I've tested both and I recon let's do 512 as that is nicely divisible by 8 and its somewhat inbetween.

**Comment by dubielzyk-expensify** [comment]
Yep agree with the gap. 

Let us know when we can re-review this :)

File: src/pages/ChatFinderPage/ChatFinderPageFooter.tsx
Comment (SzymczakJ) [reply src/pages/ChatFinderPage/ChatFinderPageFooter.tsx:6]: I guess we could think of another place to put `ReferralProgramCTA`. @Expensify/design should we also put this Referral Program in SearchRouter, or maybe should we add it somewhere else(as the space in SearchRouter is limited)

**Comment by SzymczakJ** [comment]
fixed comments and conflicts!

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (dannymcclain) [reply src/components/Search/SearchRouter/SearchRouter.tsx:156]: Beautiful! Sounds good.

**Comment by SzymczakJ** [comment]
@rayane-djouah 🙏 🙏

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/styles/variables.ts
Comment (SzymczakJ) [reply src/styles/variables.ts]: whoops, sorry for that one 😅

**Comment by SzymczakJ** [comment]
> Let us know when we can re-review this :)

This should be ready @dubielzyk-expensify

**Comment by luacmartins** [comment]
@Kicu @SzymczakJ we have conflicts now

File: src/pages/ChatFinderPage/ChatFinderPageFooter.tsx
Comment (luacmartins) [reply src/pages/ChatFinderPage/ChatFinderPageFooter.tsx:6]: I asked the product team. I'll circle back once I get an answer.

Comment (JmillsExpensify) [reply src/pages/ChatFinderPage/ChatFinderPageFooter.tsx:6]: Yeah let's not worry about this right now. We can find another place for referrals. Cc @jamesdeanexpensify

File: src/components/Search/SearchRouter/SearchRouter.tsx
Comment (Kicu) [reply src/components/Search/SearchRouter/SearchRouter.tsx:73]: No this line is correct. 

There were 2 measures done in the old `ChatFinder`. What you're thinking about is still there in `onLayout` inside SearchRouterList - https://github.com/Expensify/App/pull/49984/files#diff-6d769534d8506784b70ef165d2e6883fe38e3fde4683528be637b7f8b923ff9fR185

And this one is just the timing for performing the search of chats, copied directly from here: https://github.com/Expensify/App/blob/main/src/pages/ChatFinderPage/index.tsx#L107-L109

**Review by rayane-d — APPROVED** [review]
LGTM and tests well 🚀

**Review by luacmartins — DISMISSED** [review]
LGTM! Gonna ask the product team to test as well before we get it merged

**Comment by shawnborton** [comment]
Chiming in way late here but I'm back from OOO!

When you type text into the input, can we make sure the text uses our normal text color? Right now it's in text supporting which is wrong:
CleanShot 2024-10-15 at 17 55 21@2x

Also question for the @Expensify/design team - what height should the Recent searches be? 
CleanShot 2024-10-15 at 17 56 18@2x

I think they are 44px tall right now, but wondering if we should make them 52px to match our button/LHN row sizes?

**Comment by shawnborton** [comment]
Similar comment - when you actually have text into the input that displays above the table, we should update the text color here too:
CleanShot 2024-10-15 at 17 57 21@2x

And when it's focused, we should use the same lighter BG that the router uses as well:
CleanShot 2024-10-15 at 17 57 47@2x

**Comment by dubielzyk-expensify** [comment]
Saw that we brought this to Slack, but don't think 52px is bad at all. That being said, I think the other alignment fixes are more important than the height

**Comment by luacmartins** [comment]
@Kicu @SzymczakJ we have conflicts

**Comment by Kicu** [comment]
Conflicts solved.

I have a question, will be merging this PR soon or will we be holding up until autocomplete features are on?
Because if we want to wait we could keep the `Permission` code in this PR but still merge it to main, so that we avoid thounsands of merge commits in here

**Review by luacmartins — APPROVED** [review]

**Comment by MonilBhavsar** [comment]
Looks like a minor regression came out of this PR. Would be great if anyone can make a quick fix https://github.com/Expensify/App/issues/50977

**Comment by Kicu** [comment]
I will take a look at this @MonilBhavsar

File: src/libs/Navigation/AppNavigator/AuthScreens.tsx
Comment (shubham1206agra) [inline src/libs/Navigation/AppNavigator/AuthScreens.tsx:358]: We forgot to close the modal when the user is anonymous, which caused https://github.com/Expensify/App/issues/75498

**Pull request merged**
